src/v/cluster/partition.cc

```diff
@@ -426,6 +429,21 @@ partition::timequery(storage::timequery_config cfg) {
     cfg.max_offset = _raft->get_offset_translator_state()->to_log_offset(
       cfg.max_offset);
     result = co_await _raft->timequery(cfg);
+
+    // It's possible that during the query, our local log was GCed, and the
+    // result may not actually reflect the correct offset for this partition.
+    // If the local log doesn't look like it can serve the query anymore, query
+    // the remote log.
+    if (may_read_from_cloud && _raft->log().start_timestamp() >= cfg.time) {
```

Comment:
I think using checking against the `start_timestamp` here is the right idea. However, the problem is that currently, `disk_log_impl::start_timestamp()` does not reflect the starting "Raft" timestamp (i.e. the timestamp corresponding to `consensus::start_offset`). Instead, it's just the oldest timestamp in the log. This means that if log eviction happens during the timequery, we're prone to return the wrong result.

I can see two solutions here:
1. When `disk_log_impl` signals to `log_eviction_stm` it can provide the `start_timestamp` too. From there, we expose this "Raft start timestamp" either in `consensus` or in `disk_log_impl`.
2. implement an approximation in `disk_log_impl::start_timestamp` that's based on `_start_offset`. We can look into the index for this. This would probably be good enough.

Reply:
Thanks for helping diagnose! I'd lean towards #2, though i'm a little wary of more dragons hiding there in terms of races with GC